apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1217 @@
+
+            mTargetHeight = landShapePointer[inCellY * ESM::Land::LAND_SIZE + inCellX];
+        }
+    }
+
+    return true;
+}
+
+bool CSVRender::TerrainShapeMode::secondaryEditStartDrag (const QPoint& pos)
+{
+    return false;
+}
+
+bool CSVRender::TerrainShapeMode::primarySelectStartDrag (const QPoint& pos)
+{
+    WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+    mDragMode = InteractionType_PrimarySelect;
+    if (!hit.hit || hit.tag != 0)
+    {
+        mDragMode = InteractionType_None;
+        return false;
+    }
+    selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 0, true);
+    return false;
+}
+
+bool CSVRender::TerrainShapeMode::secondarySelectStartDrag (const QPoint& pos)
+{
+    WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+    mDragMode = InteractionType_SecondarySelect;
+    if (!hit.hit || hit.tag != 0)
+    {
+        mDragMode = InteractionType_None;
+        return false;
+    }
+    selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 1, true);
+    return false;
+}
+
+void CSVRender::TerrainShapeMode::drag (const QPoint& pos, int diffX, int diffY, double speedFactor)
+{
+    if (mDragMode == InteractionType_PrimaryEdit)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        std::string cellId = getWorldspaceWidget().getCellId (hit.worldPos);
+        mTotalDiffY += diffY;
+        if (mIsEditing == true && mShapeEditTool == 0) editTerrainShapeGrid(CSMWorld::CellCoordinates::toVertexCoords(mEditingPos), true);
+        if (mIsEditing == true && mShapeEditTool > 0) editTerrainShapeGrid(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), true);
+    }
+
+    if (mDragMode == InteractionType_PrimarySelect)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        if (hit.hit && hit.tag == 0) selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 0, true);
+    }
+
+    if (mDragMode == InteractionType_SecondarySelect)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        if (hit.hit && hit.tag == 0) selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 1, true);
+    }
+}
+
+void CSVRender::TerrainShapeMode::dragCompleted(const QPoint& pos)
+{
+    if (mDragMode == InteractionType_PrimaryEdit)
+    {
+        if (mIsEditing == true)
+        {
+            mTotalDiffY = 0;
+            mIsEditing = false;
+        }
+
+        std::sort(mAlteredCells.begin(), mAlteredCells.end());
+        std::set<CSMWorld::CellCoordinates> removeDuplicates(mAlteredCells.begin(), mAlteredCells.end());
+        mAlteredCells.assign(removeDuplicates.begin(), removeDuplicates.end());
```

Comment:
If cells are sorted there is no need to create a set. [std::unique](https://en.cppreference.com/w/cpp/algorithm/unique) is enough.

```
std::sort(mAlteredCells.begin(), mAlteredCells.end());
std::unique(mAlteredCells.begin(), mAlteredCells.end());
```

Reply:
Oh. I didn't know this. Will look into this.

Reply:
Okay. I looked into it, and remembered the reason why I did it like so. The difference is that set, assign -method changes the container size, whereas std::unique keeps the container size, leaving unspecified values in the end. Using that would therefore likely require std::erase. I'm not sure if that's still better than set-assign. I need to later loop through the vector with `for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)` and I don't want to have extra garbage in the end of the std::vector. I don't want the `for` to iterate values after the end of meaningful values.

Reply:
unique+erase helps to avoid extra allocations. I’d say it’s better for this reason in this situation. set is useful when it’s not possible to change target container.

Reply:
`std::unique` returns an iterator to the new end, so you can just call `std::vector::erase` with that iterator and the old end iterator and it should be better than creating a whole set and then putting things back into the array - the `assign` call is going to end up doing everything `erase` does and more as the old stuff is still going to be off the end of the vector's new contents.